torchaudio/prototype/functional/_ray_tracing.py

```diff
@@ -0,0 +1,156 @@
+
+    return out
+
+
+def ray_tracing(
+    room: torch.Tensor,
+    source: torch.Tensor,
+    mic_array: torch.Tensor,
+    num_rays: int,
+    e_absorption: Union[float, torch.Tensor] = 0.0,
+    scattering: Union[float, torch.Tensor] = 0.0,
+    mic_radius: float = 0.5,
+    sound_speed: float = 343.0,
+    energy_thres: float = 1e-7,
+    time_thres: float = 10.0,
+    hist_bin_size: float = 0.004,
+) -> torch.Tensor:
+    r"""Compute energy histogram via ray tracing.
+
+    The implementation is based on *pyroomacoustics* :cite:`scheibler2018pyroomacoustics`.
+
+    ``num_rays`` rays are casted uniformly in all directions from the source; when a ray intersects a wall,
+    it is reflected and part of its energy is absorbed. It is also scattered (sent directly to the microphone(s))
+    according to the ``scattering`` coefficient. When a ray is close to the microphone, its current energy is
+    recoreded in the output histogram for that given time slot.
+
+    .. devices:: CPU
+
+    .. properties:: TorchScript
+
+    Args:
+        room (torch.Tensor): The 1D Tensor to determine the room size. The shape is
+            `(D,)`, where ``D`` is 2 if room is a 2D room, or 3 if room is a 3D room. All rooms
+            are assumed to be "shoebox" rooms.
+        source (torch.Tensor): The coordinate of the sound source. Tensor with dimensions `(D)`.
+        mic_array (torch.Tensor): The coordinate of microphone array. Tensor with dimensions `(channel, D)`.
+        e_absorption (float or torch.Tensor, optional): The absorption coefficients of wall materials.
+            (Default: ``0.0``).
+            If the dtype is ``float``, the absorption coefficient is identical to all walls and
+            all frequencies.
+            If ``e_absorption`` is a 1D Tensor, the shape must be `(4,)` if the room is a 2D room,
+            representing absorption coefficients of ``"west"``, ``"east"``, ``"south"``, and
+            ``"north"`` walls, respectively.
+            Or the shape must be `(6,)` if the room is a 3D room, representing absorption coefficients
+            of ``"west"``, ``"east"``, ``"south"``, ``"north"``, ``"floor"``, and ``"ceiling"``, respectively.
+            If ``e_absorption`` is a 2D Tensor, the shape must be `(num_bands, 4)` if the room is a 2D room,
+            or `(num_bands, 6)` if the room is a 3D room. ``num_bands`` is the number of frequency bands (usually 7),
+            but you can choose other values.
+        scattering(float or torch.Tensor, optional): The scattering coefficients of wall materials.
+            (Default: ``0.0``).  The shape and type of this parameter is the same as for ``e_absorption``.
+        mic_radius(float, optional): The radius of the microphone in meters. (Default: 0.5)
+        sound_speed (float, optional): The speed of sound in meters per second. (Default: ``343.0``)
+        energy_thres (float, optional): The energy level below which we stop tracing a ray. (Default: ``1e-7``).
```

Comment:
i'd prefer "min_energy" to suggest directionality

Reply:
No strong opinion but similarly to https://github.com/pytorch/audio/pull/2850#discussion_r1041192661, consistency with pyroomacoutics might be a nice plus. LMK which one you prefer